script/add-grammar

```diff
@@ -1,124 +1,137 @@
-    warn "Command failed. Aborting."
-    exit 1
-  end
-  return if stderr.empty?
-
-  unless hide_warnings
-    stderr.each_line do |line|
-      log "  > #{line}"
-    end
-  end
-end
-
-usage = """Usage:
-  #{$0} [-q|--quiet] [--replace grammar] url
-Examples:
-  #{$0} https://github.com/Alhadis/language-roff
-  #{$0} --replace sublime-apl https://github.com/Alhadis/language-apl
-"""
-
-$replace = nil
-$verbose = true
-
-OptionParser.new do |opts|
-  opts.banner = usage
-  opts.on("-q", "--quiet", "Do not print output unless there's a failure") do
-    $verbose = false
-  end
-  opts.on("-rSUBMODULE", "--replace=SUBMODDULE", "Replace an existing grammar submodule.") do |name|
-    $replace = name
-  end
-end.parse!
-
-
-$url = ARGV[0]
-
-# No URL? Print a usage message and bail.
-unless $url
-  warn usage
-  exit 1;
-end
-
-# Exit early if docker isn't installed or running.
-log "Checking docker is installed and running"
-command('docker', 'ps')
+#!/bin/sh
+set -e
+
+usage="${0##*/} [-q|--quiet] [--replace submodule] url"
+unset replace quiet
+
+# Print non-essential feedback
+log()([ "$quiet" ] || printf '%s\n' "$@")
+
+# Print an error message
+warn()(printf '%s: %s\n' "${0##*/}" "$@")
+
+# Display a shortened help summary and bail with an error code
+bad_invocation(){
+	printf '%s\n' "$usage" >&2
+	exit 2
+}
+
+
+# Parse options
+while [ -n "$1" ]; do case $1 in
+
+	# Print an unabridged usage summary, then exit
+	-h|--help|-\?)
+		cat <<-HELP
+		Usage:
+		    $usage
+
+		Options:
+		    -q, --quiet               Do not print output unless there's a failure.
+		    -r, --replace SUBMODDULE  Replace an existing grammar submodule.
+
+		Examples:
+		    $0 https://github.com/Alhadis/language-roff
+		    $0 --replace sublime-apl https://github.com/Alhadis/language-apl
+		HELP
+		exit ;;
+
+	# Hide non-essential feedback
+	-q | --quiet)
+		quiet=1
+		break ;;
+
+	# Replace an existing submodule
+	-r* | --replace | --replace=*)
+		case $1 in
+			-r|--replace) replace=$2; shift ;; # -r [module], --replace [module]
+			-r*)          replace=${1#??}   ;; # -r[module]
+			--replace=*)  replace=${1#*=}   ;; # --replace=[module]
+		esac ;;
+
+	# Double-dash: Terminate option parsing
+	--)
+		shift
+		break ;;
+
+	# Invalid option: abort
+	--* | -?*)
+		warn 'invalid option: "%s"' "${0##*/}" "$1" >&2
+		bad_invocation ;;
+
+	# Argument not prefixed with a dash
+	*) break ;;
+
+esac; shift
+done
+
+
+# Don't proceed any further if we don't have a URL
+[ "$1" ] || bad_invocation
+
+
+# Check upfront that executables we depend on are available
+for cmd in docker git sed ruby bundle; do
+	command -v "$cmd" >/dev/null 2>&1 || {
+		warn "Required command '$cmd' not found"
+		warn 'See CONTRIBUTING.md for help on getting started: https://git.io/J0eqy'
+		exit 1
+	}
+done
+
+# Make sure Docker's running
+log 'Checking Docker is installed and running'
+docker ps >/dev/null
+
+# Make sure we're running from checkout directory
+root=`git rev-parse --git-dir`
```

Comment:
I use backticks primarily as a matter-of-taste, but since you mentioned Solaris, it's worth noting that as recently as Solaris 10, `/bin/sh` only supported the backtick form of command substitution.

Reply:
I think this is another reason to be explicit about using Bash.  I think we should stick with catering for the platforms we know, document, and have access to. I've not touched Solaris in years so have no idea what the state of things is these days and I'd rather not guess or have to try and maintain support for something that may only be used once or twice a year, if at all. This could become a limiting factor in future too so whilst we're rewriting this, we may as well keep things open for ease of development and support in future.

Reply:
Bash support isn't universal, and [Apple have deprecated it in favour of `zsh(1)`](https://www.theverge.com/2019/6/4/18651872/apple-macos-catalina-zsh-bash-shell-replacement-features). Personally, I'd rather not add Bash to an [already beastly list of dependencies](https://github.com/github/linguist/blob/HEAD/CONTRIBUTING.md#dependencies), especially when it's only needed for running a script that not every contributor is guaranteed to use.

In any case, this is less about supporting Solaris and more about cutting down on things we need to concern ourselves with. Simply by sticking to POSIX-sanctioned features and an older (yet still valid) quoting style means we don't need to worry about what shell users are running, or how recent it is. I do admit to being something of a stickler when it comes to writing portable shell-scripts, but IMHO, that can be a healthy habit to have.

If it helps, I can include a ShellCheck directive (or better yet, a `.shellcheckrc` file) that prevents this issue being flagged by future linting.

Reply:
Urgh! Good point. I forgot Apple had 🔥 Bash. 😞 

Lets do this...

> Simply by sticking to POSIX-sanctioned features and an older (yet still valid) quoting style means we don't need to worry about what shell users are running, or how recent it is.

... and this...

> I can include a ShellCheck directive (or better yet, a `.shellcheckrc` file) that prevents this issue being flagged by future linting.

We can address any issues if they crop up in future.

Reply:
Done. Ready to merge when you are.